### ISSUE TITLE
Meaning and resolution of "The downloaded file not have the expected hash!"

### ISSUE THREAD

**Issue by windmountain** [elm-tooling/elm-tooling-cli#34]
I made an app with create-elm-app, and when I try to run the tests with `elm-test` I get the following message:

```
    Failed to download:
    < https://github.com/zwilias/elm-json/releases/download/v0.2.10/elm-json-v0.2.10-x86_64-unknown-linux-musl.tar.gz
    > /home/zach/.elm/elm-tooling/elm-json/0.2.10/elm-json
    The downloaded file does not have the expected hash!
    Expected: 6ee94f04bebeb66d5ef322d76fd3dc828015571b92f1259776f716feaaec359d
    Actual:   ce0a8e237a5ac11e8dc4737c6a7adea8464cf3e0544d5dc84f7c96f74e4278dd
```

After a bit of poking around I can also try `./node_modules/.bin/elm-test`, to which I get:

```
    < https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz
    > /home/zach/.elm/elm-tooling/elm/0.19.1/elm
    The downloaded file does not have the expected hash!
    Expected: e44af52bb27f725a973478e589d990a6428e115fe1bb14f03833134d6c0f155c
    Actual:   e8b796172746fc7e2cd4edfd2e5a72d49ae38ddd89624caf16c6fa7226cb43c3
```

I would very much appreciate help figuring out the following things:

Why does `elm-test` appear to be downloading `elm-json` at runtime? (Perhaps this question is better asked over there.)
Why might I be getting these messages?
What might I try to resolve them?

**Comment by lydell** [comment]
Hi!


> Why does elm-test appear to be downloading elm-json at runtime?

For why elm-test needs elm-json, see https://github.com/rtfeldman/node-test-runner/pull/356

> Why might I be getting these messages?

Guesses:

- The whole file doesn’t get downloaded for some reason.
- Your network messes with the download.

The point of the hash check is security – you are getting something other than the expected from the Internet, so it’s not safe to extract and execute anything from it.

> What might I try to resolve them?

You can run:

```
wget https://github.com/zwilias/elm-json/releases/download/v0.2.10/elm-json-v0.2.10-x86_64-unknown-linux-musl.tar.gz
```

And then:

```
sha256sum elm-json-v0.2.10-x86_64-unknown-linux-musl.tar.gz
```

To see what hash you get.

I get `6ee94f04bebeb66d5ef322d76fd3dc828015571b92f1259776f716feaaec359d`, just like is expected in the error message.

`wc -c elm-json-v0.2.10-x86_64-unknown-linux-musl.tar.gz` says `2307188` on my computer – what does it say on yours?

**Comment by windmountain** [comment]
Indeed downloading the file with wget and running sha256sum gives me the correct (expected) hash. Continuing to dig deeper.

**Comment by lydell** [comment]
Can you try curl too, if you have it?

```
curl -#fLo elm-json.tar.gz https://github.com/zwilias/elm-json/releases/download/v0.2.10/elm-json-v0.2.10-x86_64-unknown-linux-musl.tar.gz
```

elm-tooling uses curl if it exists, otherwise wget and finally just the Node.js `https` module.

**Comment by windmountain** [comment]
Yes, curl also gives me the right thing. I've had a bit of a chance to debug and I'm seeing that the file that elm-tooling-cli downloads is about twice as large as expected. Still looking

**Comment by windmountain** [comment]
Never did get to the bottom of this, but the quick hack for me was to edit the part that shells out to curl and replace it with `x__curl` (or whatever) which will force a fallback to using wget. Once the tool was installed I couldn't figure out how to remove it and reproduce the curl problem again (I tried `rm -r ~/.elm/elm-json`?). I'll let a sleeping dog lie.

**Comment by lydell** [comment]
To remove: https://elm-tooling.github.io/elm-tooling-cli/faq/#how-do-i-uninstall

`~/.elm/elm-json/` is the place where the `elm-json` tool caches some things, not where the actual `elm-json` executable lives.

It sure is super strange that the `curl` command works when you run in yourself but fails when run from the script, and that `wget` works!

What operating system do you use?

**Comment by windmountain** [comment]
I am on Fedora 33 with curl 7.71.1

It's also seems funny that I can delete all of ~/.elm/elm-tooling and yet the `elm-test` tests run without re-downloading anything.

**Comment by lydell** [comment]
`elm-test` caches stuff in your local `elm-stuff/` folder, to avoid having to call `elm-json` at all when possible. Remove `elm-stuff/generated-code/` and it should try to download `elm-json` again.

I’ll try a Fedora live USB soon (I hope!) to see if I can reproduce.

**Comment by lydell** [comment]
I’ve tried a Fedora live USB now.

```
npm init
npm install elm-tooling
npx elm-tooling init
npx elm-tooling install
npx elm-json --version
```

All worked out without issues.

- Fedora 33 Workstation (x86_64).
- curl 7.71.1
- Node.js 12.20.1

**Comment by kawu** [comment]
Hi!  I am experiencing the same (or at least very similar) issue:
```
elm 0.19.1
    < https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz
    > /home/kuba/.elm/elm-tooling/elm/0.19.1/elm
    The downloaded file does not have the expected hash!
    Expected: e44af52bb27f725a973478e589d990a6428e115fe1bb14f03833134d6c0f155c
    Actual:   e8b796172746fc7e2cd4edfd2e5a72d49ae38ddd89624caf16c6fa7226cb43c3

elm-format 0.8.4
    < https://github.com/avh4/elm-format/releases/download/0.8.4/elm-format-0.8.4-linux-x64.tgz
    > /home/kuba/.elm/elm-tooling/elm-format/0.8.4/elm-format
    The downloaded file does not have the expected hash!
    Expected: aa6bb9d11672d8d27398746e831266c565e9837b4da6abe5d8286c2ab69ace9d
    Actual:   d987bed7f57d65c787a80a2e9040f8b13d212d45dbaefd2c135e32d3e5d4b35f

elm-json 0.2.10
    < https://github.com/zwilias/elm-json/releases/download/v0.2.10/elm-json-v0.2.10-x86_64-unknown-linux-musl.tar.gz
    > /home/kuba/.elm/elm-tooling/elm-json/0.2.10/elm-json
    The downloaded file does not have the expected hash!
    Expected: 6ee94f04bebeb66d5ef322d76fd3dc828015571b92f1259776f716feaaec359d
    Actual:   ce0a8e237a5ac11e8dc4737c6a7adea8464cf3e0544d5dc84f7c96f74e4278dd
```
The hack suggested by @windmountain (falling back to using `wget` by editing [this part](https://github.com/elm-tooling/elm-tooling-cli/blob/8c46df6a68b061779f58a651f754f118bd065e46/commands/install.ts#L438) and replacing "curl" with e.g. "x__curl") works.  Downloading the files myself with `curl` and using `sha256sum` gives the expected (correct) hash values, too.

Versions:
* curl: 7.76.1
* node: v14.17.0
* npm: 6.14.13

Any ideas on how to investigate this further?

**Comment by lydell** [comment]
Hi!

- What operating system do you use?

- Can you try making a small .js script where you execute `curl` to download for example `elm-format`, similar to how elm-tooling does it? You could start by saving to a file and running `sha256sum` in the terminal and see if you get the correct hash. If not, maybe check if the number of bytes are the same (`wc -c`)? And upload the “bad” executable here? Basically, if we can make a minimal Node.js script that causes the issue we might be able to narrow down what’s going on.

**Comment by kawu** [comment]
Thanks for your quick response!

Here's a minimal script which gives me a corrupted file with an incorrect hash: https://gist.github.com/kraklin/50d90d89ff6f45d005b7229179e859fb (thanks @kraklin for your help in getting to the bottom of this!).

Ultimately the problem turns out to be related to my local configuration of `curl` (that I completely forgot about...):
```bash
$ cat ~/.curlrc 
-w "\n"
```
So `curl` appends "\n" at the end of whatever it outputs on stdout and the resulting file is corrupted:
```
$ curl -sfL https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz --output binary-for-linux-64-bit.gz
$ sha256sum binary-for-linux-64-bit.gz
e44af52bb27f725a973478e589d990a6428e115fe1bb14f03833134d6c0f155c  binary-for-linux-64-bit.gz
$ curl -sfL https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz --output - > binary-for-linux-64-bit.gz
$ sha256sum binary-for-linux-64-bit.gz
e8b796172746fc7e2cd4edfd2e5a72d49ae38ddd89624caf16c6fa7226cb43c3  binary-for-linux-64-bit.gz
```
One solution could be to use `curl` with `-w ""` to make sure that nothing gets appended to stdout, regardless of the contents of `~/.curlrc`.  Or alternatively I can get rid of my local curl config, which also kinda solves the problem ;).

**Comment by kraklin** [comment]
I believe this [StackOverflow question and answer](https://stackoverflow.com/a/14614203) might be the reason people are having `-w "\n"` inside their `.curlrc` in the first place.

**Comment by kawu** [comment]
@kraklin exactly, this stackoverflow page is what inspired me to put `-w "\n"` inside `.curlrc` iirc.

And given that we were getting the exact same corrupted hash for the binary-for-linux-64-bit.gz file, my guess is that @windmountain also had `-w "\n"` in his local curl settings.

**Comment by lydell** [comment]
Thank you so much for investigating this!

It sounds like adding `-w ""` to override `~/.curlrc` is an easy win then.

But before I go adding that to the code – did you ever figure out why spawning `curl` from Node.js causes the issue, while “downloading the files myself with curl and using sha256sum gives the expected (correct) hash values”?

**Comment by kraklin** [comment]
There were different outcomes on whether you use `-w` or just `-o > some_file_name` So it depends on how you curl the file locally. One way the -w don't kick in and won't add the trailing `\n` character

**Comment by kawu** [comment]
I confirm, by "downloading the files myself with curl and using sha256sum gives the expected (correct) hash values" I meant that 
```
curl -sfL https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz --output binary-for-linux-64-bit.gz
```
(which I tried in the first place) works correctly (regardless of `~/.curlrc`).  The following, on the other hand:
```
curl -sfL https://github.com/elm/compiler/releases/download/0.19.1/binary-for-linux-64-bit.gz --output - > binary-for-linux-64-bit.gz
```
gives a corrupted file.

**Comment by lydell** [comment]
Ah, thanks! So it’s all about if you use stdout vs direct write to a file via `--output`.

**Comment by kraklin** [comment]
@lydell yes, exactly... sorry for my misinterpretation, I remembered it wrongly.

**Comment by lydell** [comment]
I made a pull request with improved errors and `-w ""` here: https://github.com/elm-tooling/elm-tooling-cli/pull/62

**Comment by kraklin** [comment]
Cool. I'm going to test it out with @kawu first thing tomorrow :) Thanks @lydell

**Comment by kawu** [comment]
Works perfectly. Thanks again!
